core/lib/multivm/src/versions/testonly/v26_upgrade_utils.rs

```diff
@@ -0,0 +1,289 @@
+                    .value()
+                    .0
+                    .to_vec(),
+            ),
+        ])
+        .unwrap();
+
+    let complex_upgrader = load_complex_upgrader_contract();
+    let complex_upgrader_calldata = complex_upgrader
+        .function("upgrade")
+        .unwrap()
+        .encode_input(&[
+            Token::Address(test_contract_addr),
+            Token::Bytes(test_contract_calldata),
+        ])
+        .unwrap();
+
+    let mut dependencies: HashSet<Vec<u8>> = Default::default();
+    for dep in test_contract.dependencies.iter() {
+        dependencies.insert(dep.bytecode.to_vec());
+    }
+
+    let execute = Execute {
+        contract_address: Some(COMPLEX_UPGRADER_ADDRESS),
+        calldata: complex_upgrader_calldata,
+        factory_deps: dependencies
+            .into_iter()
+            .chain(vec![beacon_proxy_bytecode])
+            .collect(),
+        value: U256::zero(),
+    };
+
+    Transaction {
+        common_data: ExecuteTransactionCommon::ProtocolUpgrade(ProtocolUpgradeTxCommonData {
+            sender: CONTRACT_FORCE_DEPLOYER_ADDRESS,
+            gas_limit: U256::from(200_000_000u32),
+            gas_per_pubdata_limit: REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_BYTE.into(),
+            ..Default::default()
+        }),
+        execute,
+        received_timestamp_ms: 0,
+        raw_bytes: None,
+    }
+}
+
+fn setup_v26_unsafe_deposits_detection<VM: TestedVm>() -> (VmTester<VM>, V26TestData) {
+    // In this test, we compare the execution of the bootloader with the predefined
+    // refunded gas and without them
+
+    let l1_chain_id = U256::from(1u32);
+
+    let test_contract = zksync_test_contracts::TestContract::bridge_test();
+    // Any random (but big) address is fine
+    let test_address = Address::from_str("abacabac00000000000000000000000000000000").unwrap();
+    // Any random (but big) address is fine
+    let l1_shared_bridge_address =
+        Address::from_str("abacabac00000000000000000000000000000001").unwrap();
+    // Any random (but big) address is fine
+    let l1_token_address = Address::from_str("abacabac00000000000000000000000000000002").unwrap();
+
+    let mut vm = VmTesterBuilder::new()
+        .with_empty_in_memory_storage()
+        .with_execution_mode(TxExecutionMode::VerifyExecute)
+        .with_custom_contracts(vec![ContractToDeploy::new(
+            test_contract.bytecode.to_vec(),
+            test_address,
+        )])
+        .with_rich_accounts(1)
+        .build::<VM>();
+
+    let system_tx = get_prepare_system_tx(
+        l1_chain_id,
+        l1_token_address,
+        l1_shared_bridge_address,
+        test_address,
+        test_contract,
+    );
+
+    vm.vm.push_transaction(system_tx);
+    let result = vm.vm.execute(InspectExecutionMode::OneTx);
+
+    assert!(!result.result.is_failed());
+
+    let l2_token_address = vm.storage.borrow_mut().read_value(&StorageKey::new(
```

Comment:
Nit: `vm.vm.read_storage(_)` would be more idiomatic / general since the new VM implementation doesn't write to `vm.storage`, but rather maintains an internal in-memory write buffer.